### ISSUE TITLE
Playable ads export seems broken with modules mode

### ISSUE THREAD

**Issue by skymen** [Scirra/Construct-3-bugs#4738]
## Problem description

Exporting a playable ad in modules mode makes the ad fail to load when in the ad networks.

## Attach a .c3p

The C3P file renamed to zip
[test bug report playable.zip](https://github.com/Scirra/Construct-3-bugs/files/6323355/test.bug.report.playable.zip)

The same project, but exported as playable so you can try directly
[New project (4).zip](https://github.com/Scirra/Construct-3-bugs/files/6323353/New.project.4.zip)

Here is the playable preview tool from facebook, but I've personally tried with Unity, Iron source and Google ads, and maybe a few more, and the same issue happened every time.
https://developers.facebook.com/tools/playable-preview/

## Steps to reproduce

1. Make a playable ad export. The project needs to make a fetch request, so you need to add a sprite, an audio file or litterally anything else
2. Open the facebook playable preview (or any other preview tool from most other ad networks) 
3. Look at black screen

## Observed result

The ad fails to load with a black screen and fails with this error:
![image](https://user-images.githubusercontent.com/11213825/114989852-25b88780-9e98-11eb-9b1f-bac2254799c5.png)
It works fine when in classic script mode

## Expected result

The ad should load.

## More details
Please note the following:
- I am about 60% sure this did not happen in previous versions of C3 and so I don't know if it's a recent C3 or Chrome change that causes this, or if by sheer luck, we never actually used modules mode in the past.
- Note that the issue is tied to CSP which I find extremely misleading because no matter how open I make the CSP manually in the index.html, it does not change the outcome
- It is kinda annoying to deal with this since C3 automatically switches to modules mode since version 220 or something, and so if I want to be able to iterate quickly without having to switch back to classic mode every time, it's a large step back :(
**Affected browsers/platforms:** I think Chrome, but Android Webview is also affected, and I think Firefox too, but I haven't tested in a while

**First affected release:** Can't tell exactly

**Comment by AshleyScirra** [comment]
As far as I can tell, the Facebook Playable Ad preview tool has been broken for some time now. The CSP imposes needless restrictions that break things.

I spent ages trying quite hard on several occasions to report the problem to Facebook and get them to do something about it, but I couldn't get anywhere and eventually gave up.

So I think the preview tool is just broken and is being left broken. You could try reporting the issue to Facebook again (good luck...). Other than that, I think it's just the tool that is broken, I believe it still works in production.

**Comment by skymen** [comment]
As I said, it's not only the Facebook preview tool that's affected. Unity's, Iron source and google's too.
And for all of the them the issue gets magically resolved by switching the script type mode, and again, even with manual CSP rules, the issue still happens which makes me thing that the error is misleading.

In all cases, if you don't plan on fixing this, at least make playable ads automatically switch back to classic mode, because in the current case, modules mode is litterally unusable.

**Comment by AshleyScirra** [comment]
Can you provide links to the exact preview tools you've been using? I've only developed the Playable Ad export against the Facebook Playable Ad specification so it's not guaranteed to actually work anywhere else if the restrictions are different, and as I said the Playable Ad preview tool appears to be pretty much just broken, but I'll check if there's anything wrong with the other tools.

FWIW in r237 I already noticed an issue where JS Modules prevented playable ads from working, and fixed it at that time. So I thought we had already sorted out modules compatibility with playable ads.

**Comment by skymen** [comment]
Hi, yes, here are a few more playable test tools you can test against:
Iron source:
https://demos.ironsrc.com/test-tool/?adUnitLoader=dapi&mode=testing
This one has a custom app where different Android WebView related bugs seem to happen. The main one being that sometimes on Android Webview, the game starts with window.innerWidth and window.innerHeight set to 0 and this crashes the game.
I made a custom script I pass my playables through that patches the runtime to fix this bug
```JS
// HTMLContent is the content of index.html as exported by C3
HTMLContent = HTMLContent.replaceAll(`this._GetCommonRuntimeOptions(opts)`, `
      await (async () => {
        async function windowShouldNotBeSizeZero() {
            console.log("Wait for window size: ", window.innerWidth, window.innerHeight);
            return new Promise(async (resolve, reject) => {
              setTimeout(async () => {
                if (window.innerWidth == 0 && window.innerHeight == 0) {
                  await windowShouldNotBeSizeZero();
                }
                resolve();
              }, 500);
            })
          }
          if (window.innerWidth == 0 && window.innerHeight == 0) {
            await windowShouldNotBeSizeZero();
          }
        return this._GetCommonRuntimeOptions(opts)
    })()`)
```
However, the custom mobile tool for Iron source only appears after you've validated everything on the web test tool, and said bug only seems to happen once every 5 or 10 tries depending on my luck.

Google: (You have to check "Select for App Campaigns")
https://h5validator.appspot.com/adwords/asset 

Unity: (I tried the IOS app)
https://apps.apple.com/us/app/ad-testing/id1463016906

TapJoy:
https://playable.tapjoy.com/

Mintegral:
https://www.mindworks-creative.com/review/

These are the most accessible ones. You don't need to validate everything, you just need to have the game show on the tool.

**Comment by AshleyScirra** [comment]
Facebook Playable Ads: our engine complies with the specifications listed here: https://www.facebook.com/business/help/412951382532338
The preview tool doesn't list any useful messages in the console and is virtually impossible to debug. As I mentioned before I believe the tool is just broken anyway.

For ironSource Playable: a Playable Ad export seems to run just fine in their test tool, so I can only assume our engine is doing everything right. It doesn't list all the requirements as met, but those appear to just be required calls to their API.
The runtime already uses 1x1 as a minimum size specifically to avoid problems with zero-sized windows, which has also come up in the past with iframes. So I don't know why a zero size would cause a problem: our engine will still make a 1x1 canvas, which should work. If they have another test tool that creates a zero-sized view, loads content in to it, *then* sets the right view size, that seems to be a bug in their tool - they should set the right view size *before* loading content in to it. So I can't say why that would be a problem for the Construct engine, but it sounds like a bug on their side anyway.

Google HTML5 validator: providing you add the meta tag they require to index.html, it passes all validation requirements.

Unity: I skipped this one - it looks like they only provide an iOS app, if I can't open up browser dev tools it's pretty much impossible to investigate anything at all.

TapJoy playable: a Playable Ad export seems to run just fine in their test tool. It complained about missing API calls, but that seems easily dealt with.

Mintegral: at first it failed to load, but after a quick look at the documentation I figured out you need a call to gameReady() to close the loading screen and show the game. Once I did that, I did discover one bug: if you have an unminified Playable Ad export using scripting, it could fail to load. I've fixed that for the next beta. It would still work though if you minified the export, which presumably you'd be doing with Playable Ad exports to meet the file size requirements. It would also work with the project file originally attached to this report, as it didn't use scripting. So once that was sorted, the game loaded and ran successfully in the preview tool. It still complained about a bunch of requirements, but they are either API call stuff that should be easily dealt with, or are frankly too poorly documented to figure out how to comply with them (e.g. it said one requirement to base64 encode everything had failed, but the playable ad export does do that, so I don't know what we're meant to do instead).

In short I'd say our engine generally does comply with all the documented restrictions and most of the preview tools seem to be working already (except for the Facebook one, which appears to be completely broken and I have no idea why, and nobody from Facebook will help me). Nobody says "Don't use JavaScript Modules", so there should not be any fundamental issue with the fact Construct uses JavaScript Modules now. If the preview tool works but it's broken elsewhere, you'd need to contact them about it, as it probably means a bug on their side, or the preview tool is incorrect in allowing things which are not actually allowed, and it's not feasible to debug undocumented limitations with no browser dev tools. If anyone can point to documented restrictions that the Construct engine does not meet, and they're straightforward to implement, I can probably do that. But as things stand, I've fixed one small bug and apart from that as best as I can tell we're holding up our end of the bargain - so please get in touch with the ad providers for any further issues beyond that.

**Comment by skymen** [comment]
Thanks a lot for your work.
I'll give everything a triple check with latest beta next week. If I still encounter issues, i'll try to contact the ad networks
For Ironsource, it is most definitely a bug on their side, as I've had the same issue with other engines, and even pure JS + three js.
For Facebook, I frankly have no idea either. And also, I agree, in many ways, playable ad documentations and tools are either buggy, unfinished, poorly documented or all of the above...

I don't see why Javascript modules would cause issues either, it's just that this is what I've encountered. Maybe it was that one bug you mentionned about unminified code. I don't minify the code for many reasons, mainly to be able to do these hotfixes I mentionned, and to be able to debug the playables once they've been deployed. The advantages of being able to dig into the exported code in a few months for whatever obscure reason far outweighs the bennefit of gaining a few KB of space.

Anyway, if I still encounter issues past the contacting of ad networks or I manage to find an issue I can prove is in C3, I'll make a new bug report.

Have a nice week
